Components: Add `DateCalendar` and `DateRangeCalendar` components

chihsuan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- Link this PR to its associated issue with an appropriate keyword: Closes, See, Follow up to, etc. -->

Related to https://github.com/Automattic/wp-calypso/pull/102989

This PR introduces new `DateCalendar` and `DateRangeCalendar` components built on the [react-day-picker](https://daypicker.dev/) library, originally implemented by @ciampo, with the following changes to make it follow the Gutenberg guidelines/patterns:

- Converted CSS modules to standard SCSS: `styles.module.scss` to `style.scss`
- Updated class names from `.calendar__*` to `.components-calendar__*`
- Moved Storybook and tests files from individual component folders to centralized `stories/` and `tests/` directory respectively
- Fixed linting errors
- Fixed unit tests (failed due to Node version differences in Intl API handling)

cc @oandregal @youknowriad 

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Current limitations: `DatePicker` (@wordpress/components) lacks date range support, while `DateRange` (Calypso) only handles ranges and requires moment.js.

Create a unified Calendar component that supports both single dates and date ranges, with flexible layout options.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Start the storybook
2. Navigate to the `DateCalendar` and `DateRangeCalendar` stories

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

<img width="1280" alt="Screenshot 2025-07-01 at 15 48 25" src="https://github.com/user-attachments/assets/8c3c534f-9f6d-4531-9a21-547afb3dd5fe" />

<img width="1273" alt="Screenshot 2025-07-01 at 15 48 33" src="https://github.com/user-attachments/assets/2daf9f53-95ed-4e68-aa19-0be32214a0b2" />


https://github.com/user-attachments/assets/af3138e6-90f3-4e48-9194-b7c11da84ae9

chihsuan committed:
Add Date Calendar and Date Range Calendar components

- Introduced `DateCalendar` and `DateRangeCalendar` components for single date and date range selection, respectively.
- Implemented localization support and accessibility features in both components.
- Added styles and utility functions for calendar rendering and interaction.
- Created tests for both components to ensure functionality and user experience.
- Updated package dependencies to include `react-day-picker` and related libraries.

chihsuan committed:
Move stories to stories directory

chihsuan committed:
Fix linting errors

chihsuan committed:
Convert styles.module.scss to style.scss and fix linting errors

chihsuan committed:
Move stories and tests to stories and test directories

chihsuan committed:
Fix lint errors/warnings

chihsuan committed:
Fix tests

chihsuan committed:
Export calendar components and update styles

- Exported `DateCalendar`, `DateRangeCalendar`, and `TZDate` from the index file.
- Updated styles in `calendar/style.scss` to remove unnecessary namespace prefixes.
- Imported calendar styles in the main `style.scss`.
- Changed story titles for `DateCalendar` and `DateRangeCalendar` to reflect the correct component hierarchy.

chihsuan committed:
Add changelog

chihsuan committed:
Update docs/manifest.json

ciampo commented on packages/components/src/calendar/date-calendar/test/index.tsx:
```diff
@@ -194,7 +233,7 @@ describe( 'DateCalendar', () => {
 		] )( '[`%s`]', ( _mode, Component ) => {
 			it( 'should select a date when a date button is clicked', async () => {
-				const user = setupUserEvent();
+				const user = await setupUserEvent();
```
The `setupUserEvent` is not asynchronous, and therefore it shouldn't require to be `await`ed — what's the reason for adding `await`? (across both test files)

chihsuan replied:
Thanks for raising this! You're right that `setupUserEvent()` itself isn’t async. However, the ESLint warning comes from the `testing-library/await-async-events` rule, which expects the return value from `setupUserEvent()` to be awaited if the methods it returns (like .click) are async. The linting tool automatically added the `await` keyword to ensure the expected behavior.

Alternatively, we could suppress the lint warning with a comment, but there are too many of these in the codebase.

<img width="922" alt="Screenshot 2025-07-04 at 11 18 20" src="https://github.com/user-attachments/assets/2be84b5e-b1bc-44b1-91f0-f0f71aa773ef" />

ciampo replied:
This looks like a false positive, since the [ESLint rule](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/await-async-events.md) is targeted at async events, while `userEvent.setup()` is a synchronous method.

After doing some digging, it looks like [tweaking the `setupUserEvent` function fixes the false-positive](https://github.com/testing-library/eslint-plugin-testing-library/issues/881).  If we tweak it to:

```
function setupUserEvent() {
	// The `advanceTimersByTime` is needed since we're using jest
	// fake timers to simulate a fixed date for tests.
	const user = userEvent.setup( { advanceTimers: jest.advanceTimersByTime } );
	return user;
}
```

we should be able to also remove all the `await`s too.

ciampo commented on packages/components/src/calendar/utils/use-controlled-value.ts:
```diff
@@ -20,6 +23,10 @@ type Props< T > = {
+ * @param root0
+ * @param root0.defaultValue
+ * @param root0.onChange
+ * @param root0.value
```
Can we avoid this code change, or otherwise document it better?

chihsuan replied:
These JSDoc block lines were added to fix ESLint warnings. I've updated parameter documentation for clarity by renaming `root0` to `props` and adjusting the parameter descriptions accordingly in be1c1f19c30c1d556c8c71ca232328ae60759af0.

```
Expected JSDoc block lines to be aligned.eslint[jsdoc/check-line-alignment](https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/check-line-alignment.md#repos-sticky-header)
Missing JSDoc @param "root0" declaration.eslintjsdoc/require-param
Missing JSDoc @param "root0.defaultValue" declaration.eslint[jsdoc/require-param](https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/require-param.md#repos-sticky-header)
Missing JSDoc @param "root0.onChange" declaration.eslintjsdoc/require-param
Missing JSDoc @param "root0.value" declaration.eslint[jsdoc/require-param](https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/require-param.md#repos-sticky-header)
```

ciampo replied:
Great, thanks!

ciampo commented on packages/components/src/calendar/style.scss:
```diff
@@ -1,19 +1,16 @@
-	--a8c-calendar-range-middle-background-color: color-mix(in srgb, #{theme.$components-color-accent} 4%, transparent);
-	--a8c-calendar-preview-border-color: color-mix(in srgb, #{theme.$components-color-accent} 16%, transparent);
+	--components-calendar-range-middle-background-color: color-mix(in srgb, $components-color-accent 4%, transparent);
+	--components-calendar-preview-border-color: color-mix(in srgb, $components-color-accent 16%, transparent);
```
Given this is a Sass file, we can refactor these internal variables to sass variables.

Also, for better namespacing, I'd prefix the variables with `wp-components`, and not only `components`.

chihsuan replied:
Sounds good! I've refactored these to sass variables and prefixed them with `wp-components`. cf6c3d5568bfc5d09001f49eace2bf44accb9e5c

ciampo replied:
🙏

ciampo commented on packages/components/src/calendar/test/date-range-calendar.tsx:
```diff
@@ -1505,10 +1515,14 @@ describe( 'DateRangeCalendar', () => {
-		tomorrow = addDays( today, 1 );
-		yesterday = subDays( today, 1 );
-		const nextWeek = addDays( today, 7 );
+		let nextWeek: Date;
+
+		beforeEach( () => {
+			today = new Date( '2024-03-15' );
+			tomorrow = addDays( today, 1 );
+			yesterday = subDays( today, 1 );
+			nextWeek = addDays( today, 7 );
+		} );
```
Why was this change necessary (re-defining the variables before each test, instead of having them as constants)?

chihsuan replied:
It's not allowed to declare these variables inside this block because they are already declared in the upper scope.

And these variables are re-assigned in some tests, so I use `beforeEach` instead of `before` to ensure tests pass consistently. An alternative approach would be to use different names for the variables; however, that would require modifying the tests directly.

```
'today' is already declared in the upper scope on line 84 column 6.eslint[@typescript-eslint/no-shadow](https://typescript-eslint.io/rules/no-shadow)
```

ciampo replied:
Got it — I guess I'd almost prefer using different variables to completely avoid the risk that the order in which tests run may affect the test result.

ciampo commented on packages/components/src/calendar/style.scss:
```diff
@@ -15,8 +15,8 @@
 	display: inline flow-root;
-	color: colors.$gray-900;
-	background-color: colors.$white;
+	color: $gray-900;
```
@mirka can you confirm that this is still the correct way to consume Sass variables? (ie. without `@use` at the top of the file and namespace (ie. `colors.$gray-900` )?

ciampo replied:
Looking at the rest of the components package, this seems indeeed still the correct way to consume Sass variables.

mirka replied:
`@use` is the correct way. Just, we haven't done the migration work yet 🙃

ciampo commented on packages/components/src/calendar/utils/use-controlled-value.ts:
```diff
@@ -0,0 +1,61 @@
+ * @param root0.onChange
+ * @param root0.value
+ */
+export function useControlledValue< T >( {
```
As a follow-up, we should try to re-use the existing `useControlledValue` hook in the package https://github.com/WordPress/gutenberg/blob/2dd93a3b957906412f6b4d8c547d8326f2ba8b78/packages/components/src/calendar/utils/use-controlled-value.ts

ciampo commented on packages/components/src/calendar/test/date-calendar.tsx:
```diff
@@ -963,7 +973,7 @@ describe( 'DateCalendar', () => {
 				<DateCalendar
 					defaultSelected={ tomorrowAtMidnightInTokyo }
-					timeZone="-02:00"
+					timeZone="Etc/GMT+2"
```
My preference would be to change the jest settings so that the tests always run in the `UTC` timezone, instead of twaking the test code — you can refer to [this comment](https://github.com/Automattic/wp-calypso/pull/103676/files#r2105315544) to see how I achieved it in the Calypso repository.

chihsuan replied:
I modified this value because Node.js 20 treats `-02:00` as an invalid timezone format. This change was necessary due to Node.js version compatibility.

```
node -p "new Intl.DateTimeFormat('en-US', { dateStyle: 'long', timeZone: '-02:00' }).format(new Date())"
[eval]:1
new Intl.DateTimeFormat('en-US', { dateStyle: 'long', timeZone: '-02:00' }).format(new Date())
^

RangeError: Invalid time zone specified: -02:00
    at new DateTimeFormat (<anonymous>)
    at [eval]:1:1
    at runScriptInThisContext (node:internal/vm:209:10)
    at node:internal/process/execution:118:14
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:101:62)
    at evalScript (node:internal/process/execution:133:3)
    at node:internal/main/eval_string:51:3

Node.js v20.14.0
```

ciampo replied:
Got it. In that case, let's leave an inline comment about it for future reference — something like `/Note: using "Etc/GMT+2" instead of "-02:00" because support for raw offsets was introduced in Node v22 (while currently the repository still targets Node v20).` — in both test files

ciampo reviewed (commented):
Thank you, @chihsuan! 

Before exporting the component from the package (thus making it a public API), I wonder if we should test the component in real-case scenario first — unless you're already using this component in WooCommerce?

_(One small note — for ease of review, it would have been better to have an initial commit only containing the raw pasted code. It would have made it easier to distinguish the actual code changes added in Gutenberg compared to the original component.)_

ciampo commented on packages/components/src/calendar/stories/date-calendar.story.tsx:
```diff
@@ -0,0 +1,220 @@
+import type { Meta, StoryObj } from '@storybook/react';
+
+const meta: Meta< typeof DateCalendar > = {
+	title: 'Components/DateCalendar',
```
```suggestion
	title: 'Components/Selection & Input/Time & Date/DateCalendar',
```

ciampo commented on packages/components/src/calendar/stories/date-range-calendar.story.tsx:
```diff
@@ -0,0 +1,230 @@
+import type { Meta, StoryObj } from '@storybook/react';
+
+const meta: Meta< typeof DateRangeCalendar > = {
+	title: 'Components/DateRangeCalendar',
```
```suggestion
	title: 'Components/Selection & Input/Time & Date/DateRangeCalendar',
```

chihsuan committed:
Update packages/components/src/calendar/stories/date-calendar.story.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chihsuan committed:
Refactor documentation in use-controlled-value.ts

- Updated parameter documentation for clarity by renaming `root0` to `props` and adjusting the parameter descriptions accordingly.

chihsuan commented:
> Thank you, @chihsuan!
> 
> Before exporting the component from the package (thus making it a public API), I wonder if we should test the component in real-case scenario first — unless you're already using this component in WooCommerce?
> 

Thanks for the detailed review! @ciampo. As discussed in Slack, we're using this component in WooCommerce Analytics and don’t have any issues with it.

> _(One small note — for ease of review, it would have been better to have an initial commit only containing the raw pasted code. It would have made it easier to distinguish the actual code changes added in Gutenberg compared to the original component.)_
 
I apologize for not clarifying this in the PR description. The initial commit does contain only the raw code without any modifications.

chihsuan committed:
Update packages/components/src/calendar/stories/date-range-calendar.story.tsx

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chihsuan committed:
Fix calendar styles

- Replaced CSS custom properties with SCSS variables for better maintainability in `style.scss`.
- Updated button and navigation dimensions, focus outlines, and background colors to utilize the new variable definitions.

chihsuan commented:
> The styles also seem to be off compared to the original copy of the component:
> 
> original copy	this branch
>  Kapture.2025-07-03.at.15.41.31.mp4 
>  Kapture.2025-07-03.at.15.42.43.mp4 
> * missing focus styles for previous/next month buttons
> * wrong focus styles for the calendar days
> * missing background color for the dates selected in a range
> * wrong dashed border color when previewing the date selection
> 
> Likely, something broke either when refactoring classnames and selectors, or when refactoring colro variables.

Good catch! @ciampo  Yes, color variables were not set correctly.

I've fixed it in cf6c3d5568bfc5d09001f49eace2bf44accb9e5c.

<img width="874" alt="Screenshot 2025-07-04 at 11 58 11" src="https://github.com/user-attachments/assets/acbca0c7-4624-4c38-b9ca-bb07553f807e" />

https://github.com/user-attachments/assets/6de889f8-7471-4e36-be3f-e075a6eeb22d

chihsuan committed:
Merge branch 'trunk' into add/calendar-component

ciampo commented on packages/components/src/calendar/style.scss:
```diff
@@ -1,52 +1,49 @@
 /* Root of the component. */
-.calendar {
+// Internal variables
+$wp-components-calendar-outline-focus: 1.5px solid $components-color-accent;
```
There should be a reusable variable (or mixin) in Gutenbegr for focus outlines

chihsuan replied:
After examining the existing styles, I can see that different components implement focus outlines in varying ways.

The existing mixins don't quite match our existing styles. 🤔 

https://github.com/WordPress/gutenberg/blob/309338a3bc19eff8010e4d7e8eb9193f4d67c138/packages/base-styles/_mixins.scss#L158-L180


I also found there is a `--wp-admin-border-width-focus` variable that I think we could use here. What do you think?

https://github.com/WordPress/gutenberg/blob/309338a3bc19eff8010e4d7e8eb9193f4d67c138/packages/base-styles/_mixins.scss#L524-L530

chihsuan replied:
Changed in 5116112c6cfac82b12b4e6cda4ff961a5430378a

ciampo replied:
Makes sense, we definitely want to use that variable

ciampo commented on packages/components/src/calendar/utils/use-controlled-value.ts:
```diff
@@ -1,7 +1,10 @@
+/**
+ * External dependencies
+ */
+import { useState, useCallback } from '@wordpress/element';
```
Probably an auto-suggest hiccup

```suggestion
import { useState, useCallback } from '@wordpress/element';
```

ciampo commented on packages/components/src/calendar/utils/use-localization-props.ts:
```diff
@@ -1,5 +1,14 @@
+/**
+ * External dependencies
+ */
+import { useMemo } from '@wordpress/element';
```
Same here

```suggestion
import { useMemo } from '@wordpress/element';
```

ciampo commented on packages/components/src/calendar/stories/date-calendar.story.tsx:
```diff
@@ -18,12 +21,18 @@ import {
+ * Internal dependencies
+ */
+import { DateCalendar, TZDate } from '..';
 import type { Meta, StoryObj } from '@storybook/react';
```
Imports from `@storybook/react` should be moved under "External dependencies"

chihsuan replied:
Done. d3ddf3c3d7bfe131054451fd61a7934778289efd

ciampo commented on packages/components/src/calendar/stories/date-range-calendar.story.tsx:
```diff
@@ -18,12 +21,18 @@ import {
+ * Internal dependencies
+ */
+import { DateRangeCalendar, TZDate } from '..';
 import type { Meta, StoryObj } from '@storybook/react';
```
Same here. re moving to external dependencies

chihsuan replied:
Done. d3ddf3c3d7bfe131054451fd61a7934778289efd

ciampo commented on packages/components/src/calendar/test/date-calendar.tsx:
```diff
@@ -14,16 +17,22 @@ import {
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
 import '@testing-library/jest-dom';
```
Potentially it may not be necessary to import `'@testing-library/jest-dom'` in Gutenberg test files.

Otherwise, same here re. moving line to external dependencies

chihsuan replied:
Removed in aa50f6483aeb152329feb7aac45221fb856a915f

ciampo commented on packages/components/src/calendar/test/date-range-calendar.tsx:
```diff
@@ -14,12 +17,18 @@ import {
 } from 'date-fns';
 import { ar } from 'date-fns/locale';
-import { useState, default as React } from 'react';
 import '@testing-library/jest-dom';
```
Potentially it may not be necessary to import `'@testing-library/jest-dom'` in Gutenberg test files.

chihsuan replied:
Yeah, it's not needed to import it. I've removed it in aa50f6483aeb152329feb7aac45221fb856a915f

ciampo commented on packages/components/src/calendar/test/date-calendar.tsx:
```diff
@@ -731,26 +904,44 @@ describe( 'DateCalendar', () => {
+					now.getUTCDate()
+				)
+			);
+			tomorrow = addDays( today, 1 );
```
Can we avoid this additional code (and any other similar UTC-related code) by changing the Jest seattings as suggested in https://github.com/WordPress/gutenberg/pull/70578/files/df7065c6b73e12b29a82d1ff6d6814897c557b8a..3053be4528a19a3f31cb368368ec2cf24318981f#r2182571977?

chihsuan replied:
Yes, thanks for pointing that out. This works perfectly. Changed in e3df37604cfdec9ed91f103c47709d1da04730ec

ciampo reviewed (commented):
> As discussed in Slack, we're using this component in WooCommerce Analytics and don’t have any issues with it.

Cool, I guess we can leave the package exports, then!

> I apologize for not clarifying this in the PR description. The initial commit does contain only the raw code without any modifications.

My bad, too — I checked, saw the `scss` file (instead of a CSS) and for some reason I assumed that wasn't the original component code. I must have gotten confused with the CSS modules setup we have in place in other repos 😅 

> Yes, color variables were not set correctly. I've fixed it

I confirm that styles look good now :)

chihsuan committed:
Update packages/components/src/calendar/utils/use-controlled-value.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chihsuan committed:
Remove unused import of '@testing-library/jest-dom' from date-calendar and date-range-calendar test files.

chihsuan committed:
Move @storybook/react  under "External dependencies"

chihsuan committed:
Update packages/components/src/calendar/utils/use-localization-props.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

chihsuan committed:
Updated the `setupUserEvent` function to return the user event directly instead of awaiting it

- Adjusted all test cases in both `date-calendar.tsx` and `date-range-calendar.tsx` to reflect this change, improving test performance and readability.

chihsuan committed:
Add note on time zone usage in DateCalendar test

- Updated the DateCalendar test to include a comment explaining the use of "Etc/GMT+2" for time zone, highlighting compatibility with Node v22.

chihsuan committed:
Create variables for preview dates instead of reusing existing variables

chihsuan committed:
Improve date handling in calendar tests

- Updated the `today`, `tomorrow`, and `yesterday` variables in `date-calendar.tsx` and `date-range-calendar.tsx` tests to ensure they are consistently set to the start of the day.
- Added a configuration to run tests in UTC timezone to avoid discrepancies based on system timezone.

chihsuan committed:
Fix conflicts

chihsuan committed:
Add note for tests

chihsuan committed:
Update calendar styles to use CSS variable for focus outline

- Changed the focus outline property in `style.scss` to utilize the CSS variable `--wp-admin-border-width-focus`, enhancing consistency and maintainability.

chihsuan commented:
Thanks, @ciampo, I've made the suggested changes. 🙇‍♂️

chihsuan committed:
Move react-day-picker imports to the external dependencies

ciampo reviewed (approved):
Thank you, @chihsuan , for the patience in addressing all of my feedback — code changes LGTM and both components test well in Storybook 🚀 

Would you be able to explore/iterate on https://github.com/WordPress/gutenberg/pull/70578#discussion_r2182557264 in a follow-up PR?

fabiankaegy commented:
I added the "Needs Dev Note" Label because I think this is a great new component we should defiantly highlight when it ships in WordPress Core  👍 :)

ciampo commented:
@fabiankaegy I'm removing the "needs dev note" label for now as I think we want to be a bit more cautious with the way we introcude the component — in #70681 I'm changing it to be a private API. Happy to add the label back in case the component is made public, though!
